preprocessors/openstreetmap/osm_service.py

```diff
@@ -174,193 +177,144 @@ def process_streets_data(OSM_data, bbox_coordinates):
+                    # for the succeeding node.
+                    bounded_nodes = add_new_node(
+                        flag, index, bounded_nodes,
+                        unbounded_nodes, bbox_coordinates)
+                else:  # If true, then it has only the preceding node.
+                    flag = False
+                    # update the "bounded_nodes" list with the estimated value
+                    # for the preceding node.
+                    bounded_nodes = add_new_node(
+                        flag, index, bounded_nodes,
+                        unbounded_nodes, bbox_coordinates)
+
+            elif i > 1:  # true if bounded_nodes has more than one
+                # node element. Only the first and the last nodes
+                # in the list are needed.
+                # So, if applicable, estimate a preceding node for the
+                # first node and a succeeding node for the last node.
+
+                # variable index gives the position of the first node element
+                # of the "bounded_nodes" list in the "unbounded_nodes" list.
+                index = unbounded_nodes.index(bounded_nodes[0])
+                if index > 0:  # if true, this first node element has
+                    # has a preceding node in the "unbounded_nodes" list.
+                    flag = False
+                    # so, update the "bounded_nodes" list with
+                    # the estimated value for the preceding node.
+                    bounded_nodes = add_new_node(
+                        flag, index, bounded_nodes,
+                        unbounded_nodes, bbox_coordinates)
+
+                # variable index gives the position of the last node element of
+                # the "bounded_nodes" list in the "unbounded_nodes" list.
+                index = unbounded_nodes.index(bounded_nodes[i - 1])
+                if index < j - 1:  # If true, this last node element has
+                    # has a succeeding node in the "unbounded_nodes" list.
+                    flag = True
+                    # so, update the "bounded_nodes" list with
+                    # the estimated value for the succeeding node.
+                    bounded_nodes = add_new_node(
+                        flag, index, bounded_nodes,
+                        unbounded_nodes, bbox_coordinates)
+    return bounded_nodes
+
+
+def add_new_node(
+        flag, index, bounded_nodes,
+        unbounded_nodes, bbox_coordinates):
+    # Note: Immediate nodes could mean the preceding node, or the
+    # succeeding node  or both.
+    # Latitude of the node element under consideration, that
+    # has its immediate node(s) outside the bounding box.
+    lat1 = unbounded_nodes[index]["lat"]
+    # Longitude of the node element under consideration, that
+    # has its immediate node(s) outside the bounding box.
+    lon1 = unbounded_nodes[index]["lon"]
+    a = (lat1, lon1)
+    if flag:  # Do for a succeeding node
```

Comment:
Please don't literally name the variable `flag`.

Reply:
Similarly, lots of duplicate code between the if and else cases here.

Reply:
The excessive logic has been largely cut down.